### ISSUE TITLE
'error': 'invalid_grant'

### ISSUE THREAD

**Issue by DragosChirila** [truffls/sign-in-with-apple-using-django#2]
Hi,

Firstly, I would like to thank you for putting this together, very informative and enlightening.

I am trying to implement Sign In with Apple in my Django project myself. I have created a new backend for Apple following your code/example and instructions from here https://python-social-auth.readthedocs.io/en/latest/backends/index.html#adding-new-backend-support .

I am pretty sure that the configuration is right, since the first call to AUTHORIZATION_URL = 'https://appleid.apple.com/auth/authorize' is working fine.

However, I am stuck with an "**invalid_grant**" error in the second call https://appleid.apple.com/auth/token. (Same as here https://forums.developer.apple.com/thread/125468 or https://forums.developer.apple.com/thread/118135)

The response in "**get_user_details**" looks like this:
```
{
	'access_token': '...',
	'token_type': 'Bearer',
	'expires_in': 3600,
	'refresh_token': '..',
	'id_token': '...',
	'error': 'invalid_grant'
}
```

Moreover, something kind of weird is happening: first time when I try the sing in, I get an user record in Django users table, something like username = **7c757fe9a3c344f085a804f19405dc17**, but no other details (empty first name, empty email, empty last name). If I delete this user and I try again, I get a different value for the username... If I try again without deleting the user, I get an error in social_core somewhere. I reckon this strange behavior is due to the above **invalid_grant** error.

Did you face this problem? Any hints will be highly appreciated.

Best regards,
Dragos

**Comment by lukaswuerzburger** [comment]
Hey @DragosChirila. First of all, thanks for bringing up this issue. We faced the **invalid_grant** error a lot of times for different reasons. Sometimes we didn't even know the real reason. 

I'm the iOS developer at Truffls. Our backend developer that implemented this is on vacation until next Monday. Let's try anyway and see how far we can get. 

So if I understood correctly, the first time you went through the signup process it worked fine without error. Only after the first time it didn't work anymore?

When you deleted your user, was that in your database or in the Apple ID account? 
- In your backend it wouldn't work a seconds time since apple gives you the user data only once per authorization.
- In the Apple ID account that would mean to remove the authorization for your app to access your user's data.

**Comment by DragosChirila** [comment]
Hey @lukaswuerzburger ,

Thank you very much for your prompt answer, it brings hope that I can get this working at some point :)

Please let me describe the problem again, I will use bullets and numbers to be easier to refer to later:

1. I am using your Python code for the Apple backend in our Django project with some small changes/additions to the configuration:

```
    name = 'apple-oauth2'
    REDIRECT_STATE = False
    AUTHORIZATION_URL = 'https://appleid.apple.com/auth/authorize' # <-- important
    ACCESS_TOKEN_URL = 'https://appleid.apple.com/auth/token'
    ACCESS_TOKEN_METHOD = 'POST' # <-- important
    SCOPE_SEPARATOR = ','
    ID_KEY = 'uid'
```
2. Apple Sign In steps:
2.1. User clicks "Apple Sing link" in our website and gets redirected to apple to enter username + password + (two factor authentication code)
2.2. the user gets "Trust this browser?" screen with options: Not now, Don't trusts and Trust. I pick Trust
2.3. the user gets a screen with a question "Do you want to continue using Sing in with Apple App with your apple ID <Email>?". I pick Continue
2.4. the user gets redirected back to the website - so there is a missing screen to ask the user to share their details! **Problem 1** ( I have tested this with several Apple accounts!)
2.5. I am logging a lot of stuff so looking inside get_user_details method I can see the value for response argument:
```
{
	'access_token': 'acf686fa2b1314201979e98675144c86a.0.nrxrz.aDuNLrF0X___VPpN-sV4ig',
	'token_type': 'Bearer', 'expires_in': 3600,
	'refresh_token': 'r134ac6eb20ef40bb88508d6801db0c2b.0.nrxrz.qiWoipose2v1AP6cJpwG8A',
	'id_token': 'eyJraWQiOiJBSURPUEsxIiwiYWxnIjoiUlMyNTYifQ.eyJpc3MiOiJodHRwczovL2FwcGxlaWQuYXBwbGUuY29tIiwiYXVkIjoiY29tLnRhbGloZWFsdGgucG9ydGFscyIsImV4cCI6MTU3NDA0MTYxNiwiaWF0IjoxNTc0MDQxMDE2LCJzdWIiOiIwMDAyMjMuYmRmNzllOTE2ZWQ2NGQ3OTgxMTU5YTMyMjU2YTYwNjQuMDAyNyIsImF0X2hhc2giOiJkbEN4bFlyb0xYcWR3Y3Ywdjg0SzJBIiwiYXV0aF90aW1lIjoxNTc0MDQxMDEzfQ.MdZK1KFqEYZXj_KU3jUooeSI6QxkxsQfUT-g8qRzg5l5azzP9rW24Y62GpBljVQb3QrVGg7Uwi-fQ5iRhvWMAvy02DESB43X21DlVh-h8fbnTWT_9PNTwUry7COZuw3-8zMGJiLkEu2fGTHWOJUrbBk8_Otorvd_Dg-zOhKmPLEAFyWFjPVC_GqkFdcavH7FZJHRvaapFauPfY20ClByz9JeffGPGLrwG-We2twPpOvyJCUPBExBQCt0Xw8VtUCI5J9RNRPSBnWvwIePf0Xxe0VLQJXysTfejEQSwu816mgukLk69mcPRMfNq0Ri8StMIGrSIKmAoG5EmLolROqvAw',
	'error': 'invalid_grant' <---- Problem 2
}
```
The id_token decoded in jwt.io :
```
Header:
{
  "kid": "AIDOPK1",
  "alg": "RS256"
}

Payload:
{
  "iss": "https://appleid.apple.com",
  "aud": "com.talihealth.portals",
  "exp": 1574041616,
  "iat": 1574041016,
  "sub": "000223.bdf79e916ed64d7981159a32256a6064.0027",
  "at_hash": "dlCxlYroLXqdwcv0v84K2A",
  "auth_time": 1574041013
}
```

Hence, the email = None.

2.6. at this point , I have a new user record in Django users with the username = 42590728930c483aaf7ec62ac13ebf94, but no other details. The user seems to be authenticated in our django website at this point. **Problem 3 - where is this random uuid coming from?**
3. If I delete the user record from step 2 (I delete the user from my django project/database, not from Apple!) and redo all the steps again, I end up with a new user with a different username = 06fc00b6bbe94c149d3a4119bf16f415 - so, this is random right?
4. If I don't delete the user and try again all the steps I end up with the following error:
```
Internal Server Error: /parent/complete/apple-oauth2/
Traceback (most recent call last):
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
    response = self._get_response(request)
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.6/contextlib.py", line 52, in inner
    return func(*args, **kwds)
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/django/views/decorators/cache.py", line 57, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/social_django/utils.py", line 49, in wrapper
    return func(request, backend, *args, **kwargs)
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/social_django/views.py", line 33, in complete
    *args, **kwargs)
  File "/var/local/envs/tali-google/lib/python3.6/site-packages/social_core/actions.py", line 71, in do_complete
    social_user.provider)
AttributeError: 'NoneType' object has no attribute 'provider'
```
5. every single time, the second request, /token one, returns the error **'error': 'invalid_grant'** in the JSON payload/response from apple.

At this point, I tried the following nodejs app example from https://auth0.com/blog/implement-sign-in-with-apple-using-auth0-extensibility/. This is not working either, the result is similar:
```
> apple_signin@1.0.0 start /var/local/tali-health/stuff/apple_signin
> node app.js

Apple Login POC listening on port 3000!
User undefined <---- PROBLEM 3
User { profile: {},
  payload: 
   { access_token: 'a489b2f4c9adb43fd86a44f85503a9dd7.0.nsst.97kSS8rqdBCpqKgxOYRIvA',
     token_type: 'Bearer',
     expires_in: 3600,
     id_token: 'eyJraWQiOiJBSURPUEsxIiwiYWxnIjoiUlMyNTYifQ.eyJpc3MiOiJodHRwczovL2FwcGxlaWQuYXBwbGUuY29tIiwiYXVkIjoiY29tLnRhbGloZWFsdGgucG9ydGFscyIsImV4cCI6MTU3NDA0MTYxNiwiaWF0IjoxNTc0MDQxMDE2LCJzdWIiOiIwMDAyMjMuYmRmNzllOTE2ZWQ2NGQ3OTgxMTU5YTMyMjU2YTYwNjQuMDAyNyIsImF0X2hhc2giOiJkbEN4bFlyb0xYcWR3Y3Ywdjg0SzJBIiwiYXV0aF90aW1lIjoxNTc0MDQxMDEzfQ.MdZK1KFqEYZXj_KU3jUooeSI6QxkxsQfUT-g8qRzg5l5azzP9rW24Y62GpBljVQb3QrVGg7Uwi-fQ5iRhvWMAvy02DESB43X21DlVh-h8fbnTWT_9PNTwUry7COZuw3-8zMGJiLkEu2fGTHWOJUrbBk8_Otorvd_Dg-zOhKmPLEAFyWFjPVC_GqkFdcavH7FZJHRvaapFauPfY20ClByz9JeffGPGLrwG-We2twPpOvyJCUPBExBQCt0Xw8VtUCI5J9RNRPSBnWvwIePf0Xxe0VLQJXysTfejEQSwu816mgukLk69mcPRMfNq0Ri8StMIGrSIKmAoG5EmLolROqvAw' } }
```

How am I testing this: I am using https://ngrok.com/ to expose my localhost on HTTPS, this way I was able to validate the domain in Apple Console.

Please let me know you want me to provide further details. Moreover, I would be happy to allow you access to my development environment to test it yourself (but I will give you that information in private).

Thanks again and looking forward to hearing from you.

Best regards,
Dragos

**Comment by DragosChirila** [comment]
The random username comes from **social_core.pipeline.user.get_username**
```
        if email_as_username and details.get('email'):
            username = details['email']
        elif details.get('username'):
            username = details['username']
        else:
            username = uuid4().hex <---------- each time it generates a new value
```

**Comment by lukaswuerzburger** [comment]
Hey @DragosChirila. It seems like you're not using the implementation for an iOS app but for a website signin, correct? We were only focusing on the native client support. If you want to implement Sign In with Apple please follow this article. We used this article too, to derive our use case from it: https://developer.okta.com/blog/2019/06/04/what-the-heck-is-sign-in-with-apple

**Comment by aamishbaloch** [comment]
Hey @DragosChirila, reading through the whole issue, I would suggest you to look into the [identifiers and keys](https://github.com/truffls/sign-in-with-apple-using-django/blob/master/identifiers-and-keys.md) on your apple's account again. (This may also resolve the Problem 1) Invalid grant doesn't actually mean something about the grant type. In our case it was the wrong bundle ID, causing this error. 

I would recommend you to verify the following:

- Bundle ID (separate for web/mobile)
- Other Keys
- Private Key
- Redirect url (in case of web) 

Without removing this error, you would not be able to move forward and I'm pretty sure once the error is removed, things will be same as mentioned in the repo. 

Don't worry about point 3 and 4 at the moment. Just focus on removing the error, you will automatically find a way afterwards. Let me know if you need me at any point.

**Comment by DragosChirila** [comment]
Hi @lukaswuerzburger , @aamishbaloch ,

Thank you very much for getting back to me with suggestions. The good news is that I got it working. So, ...

1. Two days ago I have tried our code once again and it worked ... just like that, without us changing anything in the configuration or in the code. Strange thing is that now we still receive that key 'error': 'invalid_grant' in the payload, but we also have user details (e.g. email and uid).

2. While digging further for my issue, I found this https://github.com/python-social-auth/social-core/issues/370 and I learned that a developer pushed this 12 days ago https://github.com/grazor/social-core/commit/7441c66ae67dbd79e2c1d1db1b1ca53d9a0b31aa . 

I merged his code with ours (based on yours) and in the end we've got ourselves a running AppleOuth2 backend for our Django project. Here is the final working code:
```
import json
import time
import jwt
from jwt.algorithms import RSAAlgorithm
from jwt.exceptions import PyJWTError

from social_core.backends.oauth import BaseOAuth2
from social_core.exceptions import AuthCanceled


class AppleOAuth2(BaseOAuth2):
    """ Apple authentication backend
        https://developer.apple.com/documentation/signinwithapplerestapi/authenticating_users_with_sign_in_with_apple
        https://developer.apple.com/documentation/signinwithapplerestapi/generate_and_validate_tokens
    """
    name = 'apple-oauth2'

    JWK_URL = 'https://appleid.apple.com/auth/keys'
    AUTHORIZATION_URL = 'https://appleid.apple.com/auth/authorize'
    ACCESS_TOKEN_URL = 'https://appleid.apple.com/auth/token'
    ACCESS_TOKEN_METHOD = 'POST'
    RESPONSE_MODE = None

    ID_KEY = 'sub'
    TOKEN_KEY = 'id_token'
    STATE_PARAMETER = True
    REDIRECT_STATE = False

    TOKEN_TTL_SEC = 6 * 30 * 24 * 60 * 60

    def auth_params(self, *args, **kwargs):
        """ Apple requires to set `response_mode` to `form_post`
            if `scope` parameter is passed.
        """
        params = super().auth_params(*args, **kwargs)
        if self.RESPONSE_MODE:
            params['response_mode'] = self.RESPONSE_MODE
        elif self.get_scope():
            params['response_mode'] = 'form_post'
        return params

    def get_key_and_secret(self):
        """ Return tuple with Consumer Key and Consumer Secret for current
            service provider.
            Must return (key, secret), order *must* be respected.
        """
        headers = {
            'alg': 'ES256',
            'kid': self.setting('KEY_ID')
        }

        now = int(time.time())

        payload = {
            'iss': self.setting('TEAM_ID'),
            'iat': now,
            'exp': now + self.TOKEN_TTL_SEC,
            'aud': 'https://appleid.apple.com',
            'sub': self.setting('CLIENT_ID'),
        }

        client_secret = jwt.encode(
            payload,
            self.setting('PRIVATE_KEY'),
            algorithm='ES256',
            headers=headers
        ).decode('utf-8')

        return self.setting('CLIENT_ID'), client_secret

    def get_apple_jwk(self):
        keys = self.get_json(url=self.JWK_URL).get('keys')
        if not isinstance(keys, list) or not keys:
            raise AuthCanceled('Invalid jwk response')

        return json.dumps(keys.pop())

    def decode_id_token(self, id_token):
        """ Decode and validate JWT token from apple and return payload
            including user data.
        """
        if not id_token:
            raise AuthCanceled(f'Missing {self.TOKEN_KEY} parameter')

        public_key = RSAAlgorithm.from_jwk(self.get_apple_jwk())
        try:
            decoded = jwt.decode(
                id_token,
                key=public_key,
                audience=self.setting('CLIENT_ID'),
                algorithm='RS256',
            )
        except PyJWTError:
            raise AuthCanceled('Token validation failed')

        return decoded

    def get_user_details(self, response):
        name = response.get('name') or {}
        fullname, first_name, last_name = self.get_user_names(
            fullname='',
            first_name=name.get('firstName', ''),
            last_name=name.get('lastName', ''),
        )

        email = response.get('email', '')
        apple_id = response.get(self.ID_KEY, '')

        user_details = {
            'username': apple_id,
            'first_name': first_name,
            'last_name': last_name,
            'email': email,
        }

        return user_details

    def do_auth(self, access_token, response, *args, **kwargs):
        jwt_string = (
            kwargs.pop('jwt_string', None) or response.get(self.TOKEN_KEY)
        )
        if not jwt_string:
            raise AuthCanceled(f'Missing {self.TOKEN_KEY} parameter')

        decoded_data = self.decode_id_token(jwt_string)
        return super().do_auth(
            access_token, response=decoded_data, *args, **kwargs
        )

```
 
Looking forward to having that pull request accepted and integrated in python-social library. Until then, we will be using the above.

Cheers,
Dragos

**Comment by Johnywhisky** [comment]
Hi there, I faced on this error during following you
`TypeError: __init__() missing 1 required positional argument: 'strategy'`
plz someone let me know how to solve it :(
